ReconcileCompatibility should not parse the current version

mikechengwei opened the pull request:
Fix #4233

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

thbkrkr commented on pkg/controller/common/annotation/controller_version.go:
```diff
@@ -73,6 +73,11 @@ func ReconcileCompatibility(ctx context.Context, client k8s.Client, obj ctrlclie
+	if err != nil {
+		return false, errors.Wrap(err, "Error parsing controller version")
+	}
+
```
Like @barkbay proposed in the inital issue, I think we should only attempt to parse the controller version before we set it in the resource annotation (before L45 in UpdateControllerVersion):

https://github.com/elastic/cloud-on-k8s/blob/1f8464a90f495368a00c8264cbaf5148ab645964/pkg/controller/common/annotation/controller_version.go#L45

It will be good to rename the variable `version` in `controllerVersion` to avoid a conflict with the `version` package used to call the function `Parse`.

thbkrkr reviewed (commented):
Thanks for the contribution! 
I left a comment to move the version validity check with a parsing.
Please also update the unit tests `TestAnnotationUpdated` and `TestAnnotationCreated` with valid versions and add a new unit test `TestInvalidAnnotation` after these two.

thbkrkr reviewed (changes requested):
Please also update the unit tests `TestAnnotationUpdated` and `TestAnnotationCreated` with valid versions and add a new unit test `TestInvalidAnnotation` after these two.

mikechengwei commented:
Thank you for your guidance, I have optimized it, please help me review it. @thbkrkr

thbkrkr reviewed (commented):
It remains a minor correction for the godoc and it will be perfect.

thbkrkr commented on pkg/controller/common/annotation/controller_version_test.go:
```diff
@@ -275,6 +275,23 @@ func TestNewerAnnotation(t *testing.T) {
 	assert.True(t, compat)
 }
 
+// Test TestInvalidAnnotation check if the annotation is valid.
```
Uh, there are other comments for the Godoc who are terrible in this file. It would be nice to fix them in the process.

```diff
diff --git a/pkg/controller/common/annotation/controller_version_test.go b/pkg/controller/common/annotation/controller_version_test.go
index 318d13c63..a98210117 100644
--- a/pkg/controller/common/annotation/controller_version_test.go
+++ b/pkg/controller/common/annotation/controller_version_test.go
@@ -108,7 +108,7 @@ func TestCheckCompatibility(t *testing.T) {
        }
 }
 
-// Test UpdateControllerVersion updates annotation if there is an older version
+// TestAnnotationUpdated tests that an annotation is updated if there is an older version
 func TestAnnotationUpdated(t *testing.T) {
        kibana := kbv1.Kibana{
                ObjectMeta: metav1.ObjectMeta{
@@ -126,7 +126,7 @@ func TestAnnotationUpdated(t *testing.T) {
        require.Equal(t, obj.GetAnnotations()[ControllerVersionAnnotation], "1.7.0")
 }
 
-// Test UpdateControllerVersion creates an annotation even if there are no current annotations
+// TestAnnotationCreated tests that an annotation is created even if there are no current annotations
 func TestAnnotationCreated(t *testing.T) {
        kibana := kbv1.Kibana{
                ObjectMeta: metav1.ObjectMeta{
@@ -220,7 +220,6 @@ func TestMissingAnnotationNewObject(t *testing.T) {
        assert.Equal(t, MinCompatibleControllerVersion, es.Annotations[ControllerVersionAnnotation])
 }
 
-//
 func TestSameAnnotation(t *testing.T) {
        es := &esv1.Elasticsearch{
                ObjectMeta: metav1.ObjectMeta{
@@ -275,7 +274,7 @@ func TestNewerAnnotation(t *testing.T) {
        assert.True(t, compat)
 }
```

mikechengwei replied:
According to your suggestion, I have optimized it.

thbkrkr commented:
Jenkins test this please

mikechengwei commented:
/test

mikechengwei commented:
how to kick off a build manually? @thbkrkr I'm first submit.

mikechengwei commented:
![image](https://user-images.githubusercontent.com/13257127/129338630-dfebb160-f78a-498d-a55f-6ef19d8c1a37.png)

thbkrkr commented:
> how to kick off a build manually? @thbkrkr I'm first submit.

You can't for security reason but I triggered it with this [comment](https://github.com/elastic/cloud-on-k8s/pull/4748#issuecomment-898328672).

I ran all unit tests locally and there is another test tp update in the `enterprisesearch` package that used an invalid controller version:

```diff
diff --git a/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go b/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go
index 2bf0ae4f0..b1415b220 100644
--- a/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go
+++ b/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go
@@ -77,7 +77,7 @@ func TestReconcileEnterpriseSearch_Reconcile_SetControllerVersion(t *testing.T)
                Parameters: operator.Parameters{
                        OperatorInfo: about.OperatorInfo{
                                BuildInfo: about.BuildInfo{
-                                       Version: "operator-version",
+                                       Version: "1.7.0",
                                },
                        },
                },
@@ -89,7 +89,7 @@ func TestReconcileEnterpriseSearch_Reconcile_SetControllerVersion(t *testing.T)
        var updated entv1.EnterpriseSearch
        err = r.Client.Get(context.Background(), k8s.ExtractNamespacedName(&sample), &updated)
        require.NoError(t, err)
-       require.Equal(t, map[string]string{annotation.ControllerVersionAnnotation: "operator-version"}, updated.Annotations)
+       require.Equal(t, map[string]string{annotation.ControllerVersionAnnotation: "1.7.0"}, updated.Annotations)
```

mikechengwei commented:
jenkins test this please

mikechengwei commented:
> > how to kick off a build manually? @thbkrkr I'm first submit.
> 
> You can't for security reason but I triggered it with this [comment](https://github.com/elastic/cloud-on-k8s/pull/4748#issuecomment-898328672).
> 
> I ran all unit tests locally and there is another test tp update in the `enterprisesearch` package that used an invalid controller version:
> 
> ```diff
> diff --git a/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go b/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go
> index 2bf0ae4f0..b1415b220 100644
> --- a/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go
> +++ b/pkg/controller/enterprisesearch/enterprisesearch_controller_test.go
> @@ -77,7 +77,7 @@ func TestReconcileEnterpriseSearch_Reconcile_SetControllerVersion(t *testing.T)
>                 Parameters: operator.Parameters{
>                         OperatorInfo: about.OperatorInfo{
>                                 BuildInfo: about.BuildInfo{
> -                                       Version: "operator-version",
> +                                       Version: "1.7.0",
>                                 },
>                         },
>                 },
> @@ -89,7 +89,7 @@ func TestReconcileEnterpriseSearch_Reconcile_SetControllerVersion(t *testing.T)
>         var updated entv1.EnterpriseSearch
>         err = r.Client.Get(context.Background(), k8s.ExtractNamespacedName(&sample), &updated)
>         require.NoError(t, err)
> -       require.Equal(t, map[string]string{annotation.ControllerVersionAnnotation: "operator-version"}, updated.Annotations)
> +       require.Equal(t, map[string]string{annotation.ControllerVersionAnnotation: "1.7.0"}, updated.Annotations)
> ```

I have fix this problem.

thbkrkr commented:
Jenkins test this please

thbkrkr commented:
Jenkins test this please

thbkrkr reviewed (approved):
LGTM
